### PR TITLE
 [JENKINS-49114] evaluate result and TestResultAction late

### PR DESCRIPTION
**Pull request by mwinter69**
Do not get build result and TestResultAction when the BuildData is initialized. Instead check if it is available when creating the json and get it one time when it is set.

This might fix the problem JENKINS-49114

### CODE REVIEW

File: src/main/java/jenkins/plugins/logstash/persistence/BuildData.java [hunk 1]
```diff
@@ -256,19 +256,27 @@ private void initData(Run<?, ?> build, Date currentTime) {
     description = build.getDescription();
     url = build.getUrl();
     buildNum = build.getNumber();
+    buildDuration = currentTime.getTime() - build.getStartTimeInMillis();
+    timestamp = DATE_FORMATTER.format(build.getTimestamp().getTime());
+  }
 
+  private void updateResult()
```
Comment (jakub-bochenski) [inline src/main/java/jenkins/plugins/logstash/persistence/BuildData.java]: why not move this to `getTestResults()`?

**Comment by jakub-bochenski** [comment]
I don't like the `BuildData` growing so much.
Maybe we could come up with some better way of handling `Action`s.

I think a more generic approach might be useful to allow sending data from other action types (e.g. warnings)

File: src/main/java/jenkins/plugins/logstash/persistence/BuildData.java (hunk 1 above)
Comment (mwinter69) [reply src/main/java/jenkins/plugins/logstash/persistence/BuildData.java]: getTestResults() is just a getter. And we also have getResult(). The updateResult() is updating both.
And honestly BuildData should not really be public api and used outside of logstash. This is an implementation detail.

**Comment by mwinter69** [comment]
The current implementation is anyway problematic. There might more than 1 AbstractTestResultAction added to a build but only the first is considered. Changing this would require a change in the json output or at least an extension. But this should not be part of this PR.

File: src/main/java/jenkins/plugins/logstash/persistence/BuildData.java (hunk 1 above)
Comment (jakub-bochenski) [reply src/main/java/jenkins/plugins/logstash/persistence/BuildData.java]: OK, can we do something else to remove the need for a explicit call to `buildData.toJson();` in  tests?

**Comment by jakub-bochenski** [comment]
My main issue is: while currently BuildData is not immutable by definition it doesn't really change much after constructor call.  Except for `initData()` it's a dumb DTO.

Now we are going to introduce changing state -- this makes the object one notch more complex. 
Maybe it would be better to leave "static" data in BuildData and have a separate mechanism to handle lazy-evaluated data?

**Comment by mwinter69** [comment]
Consider that currently the json is just a serialization of BuildData with gson. If you want to remove the non static data you're changing the json layout which is an incompatible change,
If you want to keep it as it is then you need modify the json before it is converted to a string.
You can also keep result and testresult as it is in BuildData and add an additional element, but then you duplicate data.

**Comment by jakub-bochenski** [comment]
@mwinter69 I've resolved the conflicts, can you please check if I did it right?

I still think this part of code needs a major refactoring, but I don't want it to block this improvement

**Comment by mwinter69** [comment]
some minor things, fixed them

**Pull request merged**